File: client/python/accessor.py
import socket


class Accessor:

    SPACE_REPLACEMENT = "~`\t"

    def __init__(self, host="localhost", port=8000) -> None:
        self.host = host
        self.port = port

        self.init_connection()

    def init_connection(self):
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.socket.connect((self.host, self.port))

    def __send(self, s):
        self.socket.sendall(bytes((s + "\n"), "utf-8"))

    def __receive(self):
        return self.socket.recv(1024).decode("utf-8")[:-1]

    def get(self, key):
        self.__send(f"GET {key}")
        resp = self.__receive()
        if resp[0] == "P":
            _, t, val = resp.split(" ")
            return Accessor.resolve_type(t, val)
        raise Accessor.KeyException(resp[2:])

    def put(self, key, obj):
        t, val = Accessor.resolve_object(obj)
        self.__send(f"PUT {key} {t} {val}")
        resp = self.__receive()
        if resp[0] != "P":
            raise Accessor.TypeException(resp[2:])

    def set(self, key, obj):
        _, val = Accessor.resolve_object(obj)
        self.__send(f"SET {key} {val}")
        resp = self.__receive()
        if resp[0] != "P":
            raise Accessor.ServerException(
                "Something went wrong with the server as this command should never fail."
            )

    def update(self, key, obj):
        _, val = Accessor.resolve_object(obj)
        self.__send(f"UPDATE {key} {val}")
        resp = self.__receive()
        if resp[0] != "P":
            arr = resp.split(" ")
            if arr[1] == "key":
                raise Accessor.KeyException(resp[2:])
            elif arr[1] == "value":
                raise Accessor.TypeException(resp[2:])

    def delete(self, key):
        self.__send(f"DELETE {key}")
        resp = self.__receive()
        if resp[0] != "P":
            raise Accessor.KeyException(resp[2:])

    def doc(self, command):
        self.__send(f"DOC {command}")
        resp = self.__receive()
        if resp[0] != "P":
            raise Accessor.ServerException(resp[2:])
        return resp[2:]

    def keys(self):
        self.__send("KEYS")
        resp = self.__receive()
        if resp[0] != "P":
            raise Accessor.ServerException(
                "Something went wrong with the server as this command should never fail."
            )
        return resp[3:-1].split(", ")

    def clear(self):
        self.__send("CLEAR")
        resp = self.__receive()
        if resp[0] != "P":
            raise Accessor.ServerException(resp[2:])

    def __getitem__(self, key):
        return self.get(key)

    def __setitem__(self, key, val):
        self.set(key, val)

    def __str__(self):
        self.__send("DISPLAY")
        resp = self.__receive()
        if resp[0] != "P":
            raise Accessor.ServerException(
                "Something went wrong with the server as this command should never fail."
            )
        return resp[2:].replace(Accessor.SPACE_REPLACEMENT, " ")

    @staticmethod
    def resolve_object(obj):
        t = type(obj).__name__
        val = str(obj)
        if t == "int":
            if abs(obj) < 2147483648 / 10:
                return "integer", val
            else:
                return "long", val
        elif t == "float":
            return "double", val
        elif t == "str":
            return "string", val.replace(" ", Accessor.SPACE_REPLACEMENT)
        elif t == "bool":
            return "boolean", val.lower()
        return t, val

    @staticmethod
    def resolve_type(t, val):
        if t in {"int", "integer", "short", "long"}:
            return int(val)
        elif t in {"float", "double"}:
            return float(val)
        elif t in {"bool", "boolean"}:
            return val.title() == "True"
        elif t in {"str", "string", "chr", "char"}:
            return val.replace(Accessor.SPACE_REPLACEMENT, " ")
        return val

    def close(self):
        self.__send("EXIT")
        self.__receive()
        self.socket.shutdown(socket.SHUT_RDWR)
        self.socket.close()

    class ServerException(Exception):
        pass

    class KeyException(Exception):
        pass

    class TypeException(Exception):
        pass


if __name__ == "__main__":
    import sys

    port = 8000 if len(sys.argv) == 1 else int(sys.argv[1])

    accessor = Accessor(port=port)

    x = accessor.get("x")
    x += 1
    accessor.update("x", x)

    accessor.put("hi", 1.2)
    accessor.put("test", True)
    accessor.put("stringTest", "hisdflksdjf")
    accessor.put("stringTest2", "SPACE HERE")

    print(accessor.get("stringTest2"))
    print(accessor)

    accessor.delete("stringTest2")

    print(accessor.doc("doc"))

    accessor["y"] = 2.5
    print(accessor["y"])

    for key in accessor.keys():
        print(key + " ", end="")
    print()

    accessor.close()

File: client/js/accessor.mjs
import net from 'net';

class Accessor {

    static SPACE_REPLACEMENT = "~`\t";

    constructor(host="localhost", port=8000) {
        this.host = host;
        this.port = port;
        this._receiveListeners = [];
        const _this = this;
        // this.obj = new Proxy({}, {
        //     get: async (proxy, key) => await _this.get(key),
        //     set: async (proxy, key, value) => {
        //         return await _this.set(key, value) === undefined;
        //     }
        // });
        return this.initConnection();
    }

    async initConnection() {
        const socket = new net.Socket();
        const _this = this;
        this.socket = await new Promise((res, rej) => {
            socket.connect(this.port, this.host, () => {
                socket.on("data", (data) => {
                    let str = String.fromCharCode(...data)
                    str = str.substring(0, str.length - 1);
                    for (const listener of _this._receiveListeners)
                        listener(str);
                    _this._receiveListeners = [];
                });
                res(socket);
            });
            socket.on("error", err => rej(err));
        });;
        return this;
    }

    _send(s) {
        this.socket.write(s + "\n");
    }

    async _receive() {
        const _this = this;
        return await new Promise((res, rej) => {
            _this._receiveListeners.push(res);
        });
    }

    async get(key) {
        this._send(`GET ${key}`);
        const resp = await this._receive();
        if (resp[0] === "P") {
            const [_, t, val] = resp.split(" ");
            return Accessor.resolveType(t, val);
        }
        throw new Accessor.KeyException(resp.substring(2));
    }

    async put(key, obj) {
        const [t, val] = Accessor.resolveObject(obj);
        this._send(`PUT ${key} ${t} ${val}`);
        const resp = await this._receive();
        if (resp[0] !== "P")
            throw new Accessor.TypeException(resp.substring(2));
    }

    async set(key, obj) {
        const [_, val] = Accessor.resolveObject(obj);
        this._send(`SET ${key} ${val}`);
        const resp = await this._receive();
        if (resp[0] !== "P")
            throw new Accessor.ServerException("Something went wrong with the server as this command should never fail.");
    }

    async update(key, obj) {
        const [_, val] = Accessor.resolveObject(obj);
        this._send(`UPDATE ${key} ${val}`);
        const resp = await this._receive();
        if (resp[0] !== "P") {
            const arr = resp.split(" ");
            if (arr[1] === "key") {
                throw new Accessor.KeyException(resp.substring(2));
            } else {
                throw new Accessor.TypeException(resp.substring(2));
            }
        }
    }

    async delete(key) {
        this._send(`DELETE ${key}`);
        const resp = await this._receive();
        if (resp[0] !== "P")
            throw new Accessor.KeyException(resp.substring(2));
    }

    async doc(command) {
        this._send(`DOC ${command}`);
        const resp = await this._receive();
        if (resp[0] !== "P")
            throw new Accessor.ServerException(resp.substring(2));
        return resp.substring(2);
    }

    async keys() {
        this._send("KEYS");
        const resp = await this._receive();
        if (resp[0] !== "P")
            throw new Accessor.ServerException("Something went wrong with the server as this command should never fail.");
        return resp.substring(3, resp.length - 1).split(", ");
    }

    async clear() {
        this._send("CLEAR");
        const resp = await this._receive();
        if (resp[0] !== "P")
            throw new Accessor.ServerException(resp.substring(2));
    }

    async getString() {
        this._send("DISPLAY");
        const resp = await this._receive();
        if (resp[0] !== "P")
            throw new Accessor.ServerException("Something went wrong with the server as this command should never fail.");
        return resp.substring(2).replaceAll(Accessor.SPACE_REPLACEMENT, " ")
    }

    static resolveObject(obj) {
        const t = typeof obj;
        const val = obj.toString();
        switch (t) {
            case "number":
                if (obj % 1 == 0) {
                    if (Math.abs(obj) < 2147483648 / 10)
                        return ["integer", val];
                    else
                        return ["long", val];
                } else
                    return ["double", val];
            case "string":
                return [t, val.replaceAll(" ", Accessor.SPACE_REPLACEMENT)];
            case "boolean":
                return [t, val];
            default:
                return [t, val];
        }
    }

    static resolveType(t, val) {
        switch (t) {
            case "int":
            case "integer":
            case "short":
            case "long":
                return parseInt(val);
            case "float":
            case "double":
                return parseFloat(val);
            case "bool":
            case "boolean":
                return val.toLowerCase() === "true";
            case "str":
            case "string":
            case "chr":
            case "char":
                return val.replaceAll(Accessor.SPACE_REPLACEMENT, " ");
        }
        return val;
    }        

    close() {
        this._send("EXIT");
        this.socket.destroy();
    }

    static ServerException = class extends Error {
        constructor(...params) {
            super(...params);
        }
    };

    static KeyException = class extends Accessor.ServerException {
        constructor(...params) {
            super(...params);
        }
    };

    static TypeException = class extends Accessor.ServerException {
        constructor(...params) {
            super(...params);
        }
    };
}


export default Accessor;